### DISCUSSION TITLE
how to scan the use of weak encryption suites

### DISCUSSION THREAD

**Discussion by PolarPeak** [projectdiscovery/tlsx#305] [Q&A]
I used TLSx -u xxx.com -ce for detection, and when I saw the returned content, I was unsure if it listed all the encryption suites used by my website or enumerated and displayed the cipher suites supported by my HTTPS website.
![image](https://github.com/projectdiscovery/tlsx/assets/44831329/22037271-86dc-48ec-be4d-03ba87d1a3fc)

**Comment by ehsandeep** [comment]
@PolarPeak, you can choose specific type of ciphers to scan for, in this case you are looking for weak.

```console
tlsx -h ct

Usage:
  tlsx [flags]

Flags:
   -ct, -cipher-type value  ciphers types to enumerate (all/secure/insecure/weak) (default all)
```

```
echo google.com | tlsx -ce -ct weak
```

**Reply by PolarPeak** [reply]
When I executed "./tlsx -u xxx -ce -ct weak," I did not get the desired result. I was unable to find "TLS_RSA_WITH_RC4_128_SHA."
<img width="699" alt="1689133847544" src="https://github.com/projectdiscovery/tlsx/assets/44831329/37ed90bf-bab3-4ca0-9e6e-1dc0d2424e37">

When I execute ./tlsx -u xxx -ce -j, I see related content in insecure, I don’t understand
![image](https://github.com/projectdiscovery/tlsx/assets/44831329/b340e0e1-5094-4f1e-b2ae-a53b58f6f443)

**Comment by bad-antics** [comment]
The `-ce` (cipher enumeration) output shows **all cipher suites the server supports** — it actively negotiates each one to test if the server accepts it. So yes, it's enumerating what your server supports, not just what was used in the connection.

**To identify weak ciphers from the output:**

```bash
# List all ciphers and flag weak ones
echo "xxx.com" | tlsx -ce -json | jq '.cipher_enum[]' | while read cipher; do
    cipher=$(echo $cipher | tr -d '"')
    if echo "$cipher" | grep -qiE 'RC4|DES|NULL|EXPORT|anon|MD5'; then
        echo "WEAK: $cipher"
    elif echo "$cipher" | grep -qiE 'CBC'; then
        echo "MODERATE: $cipher (CBC mode - vulnerable to BEAST/Lucky13)"
    else
        echo "OK: $cipher"
    fi
done
```

**Quick weak cipher check:**
```bash
# Just show weak ciphers
echo "xxx.com" | tlsx -ce -json | jq '[.cipher_enum[] | select(test("RC4|DES|NULL|EXPORT|anon"; "i"))]'
```

**What's considered weak in 2025+:**
- ❌ `RC4` — broken
- ❌ `DES` / `3DES` — SWEET32 attack
- ❌ `NULL` — no encryption
- ❌ `EXPORT` — deliberately weakened
- ❌ `anon` — no authentication
- ⚠️ `CBC` — vulnerable to padding oracle attacks
- ✅ `ECDHE_*_GCM_*` — modern and secure
- ✅ `CHACHA20_POLY1305` — modern and secure

**Best practice audit:**
```bash
echo "xxx.com" | tlsx -ce -tv -json | jq '{host: .host, tls_versions: .tls_version, total_ciphers: (.cipher_enum | length), weak: [.cipher_enum[] | select(test("RC4|DES|NULL|EXPORT|anon"; "i"))]}'
```
